### ISSUE TITLE
Close the CUDA indexing and layout gap (11 operations)

### ISSUE THREAD

**Issue by xupremix** [xupremix/incin#88]
## Problem

CUDA cannot index and cannot reshape. Eleven catalog operations covering gather/scatter and layout have no CUDA executor:

- **Indexing (5):** `embedding`, `gather`, `scatter`, `index_select`, `diag`
- **Shape and layout (6):** `pad`, `repeat`, `tril`, `triu`, `unfold`, `pixel_shuffle`

`embedding` is the first layer of every language model. `gather`, `scatter`, and `index_select` are how a mixture-of-experts layer permutes tokens to experts and back. `tril` and `triu` are how a causal attention mask is built. None of them run on CUDA today, so the operations that make a transformer a transformer are precisely the ones that force a fallback.

`gather`, `scatter`, and `index_select` are also the operations where a wrong index is a memory-safety question rather than a numerical one, so they need bounds behavior settled rather than assumed.

## Proposed implementation

- Implement `embedding` forward and backward. The backward is a scatter-add over a potentially repeated index set, so it needs a defined accumulation order or an explicitly documented nondeterminism, not silent atomics.
- Implement `gather`, `scatter`, and `index_select` with bounds checking that fails the launch with the typed error contract rather than reading out of range, and record the out-of-range behavior in `docs/OPERATION_SEMANTICS.md`.
- Implement `pad`, `repeat`, `tril`, `triu`, and `pixel_shuffle` as index-mapping kernels sharing one strided-copy skeleton.
- Implement `unfold` against the same descriptor the convolution path uses rather than a second geometry derivation.
- Keep `diag` consistent between its extract and construct forms.

## Acceptance criteria

- All 11 operations execute on CUDA and match the CPU reference.
- An out-of-range index produces the documented typed error on CUDA, matching CPU, and never reads outside the allocation. Covered by a test that passes an out-of-range index deliberately.
- `embedding` backward accumulates repeated indices correctly, and its determinism guarantee is documented and tested.
- Gradients match the CPU reference for every row advertised as training-capable.
- Non-contiguous, wrong-dtype, and wrong-device inputs remain rejected before launch.

## Proposed API and capability-row change

No new public API. The eleven operations exist as `Tensor` methods with CPU executors; the CUDA capability table gains rows and the executors that obligates.

```rust
use incin::prelude::*;
type Backend = DefaultBackend;   // with `cuda`, the CUDA backend

// The first layer of every language model:
let emb: Embedding<s![50_257, 512], Backend> = Embedding::build(())?;
let h = emb.forward(token_ids)?;              // today on CUDA: capability refusal

// How a causal mask is built:
let mask = Tensor::<s![1024, 1024], Backend, bool>::ones(())?.triu(1)?;  // refusal
let scores = scores.masked_fill(&mask, f32::NEG_INFINITY)?;

// How a mixture-of-experts layer permutes tokens to experts and back:
let grouped = tokens.index_select(&order, 0)?;   // refusal
```

The operations that make a transformer a transformer are precisely the ones that force a CPU fallback today.

## Shape and type-system invariants

| Operation | Shape rule | Safety or contract note |
|---|---|---|
| `embedding` | `[..]` i64 indices, `[V, D]` table `-> [.., D]` | Backward is a **scatter-add over a repeated index set** |
| `gather(index, dim)` | index shape drives output | **Bounds** |
| `scatter(index, src, dim)` | shape-preserving on `self` | **Bounds**; overwrite semantics (`scatter_add` is #103) |
| `index_select(index, dim)` | `[.., D, ..] -> [.., N, ..]` | **Bounds** |
| `diag` | `[N] -> [N, N]` or `[N, N] -> [N]` | Extract and construct forms must agree |
| `pad` | extent arithmetic per axis | |
| `repeat` | `[.., D, ..] -> [.., D*r, ..]` | |
| `tril`, `triu` | shape-preserving, `diagonal` attribute | |
| `unfold` | convolution geometry | Must use the **same descriptor** as `conv2d`, not a second derivation |
| `pixel_shuffle` | `[N, C*r*r, H, W] -> [N, C, H*r, W*r]` | `C % (r*r) == 0` |

**The three bounds rows are a memory-safety question, not a numerical one.** An out-of-range index in `gather`, `scatter`, or `index_select` must fail the launch with the typed error contract and must never read outside the allocation. This is why the issue carries `area:safety`. The behavior needs a test that passes an out-of-range index deliberately, on device, and it must be documented in `docs/OPERATION_SEMANTICS.md` rather than left to whatever the kernel happens to do.

**`embedding` backward needs a determinism decision.** Accumulating gradients for a token that appears many times in a batch is order-dependent in floating point. Atomics are fast and nondeterministic; a sorted-segment reduction is deterministic and slower. Whichever is chosen, the guarantee is part of the row and must match what `scatter_add` decides in #103, or the two disagree about the same operation shape.

## How this interoperates

- **#101** cannot run on CUDA without `embedding`, `tril`/`triu`, and `masked_fill`. This is that issue's hard backend dependency.
- **#102 and #103** need `gather`, `scatter`, and `index_select` on device for routing; #103 adds `scatter_add` with the accumulate semantics `embedding` backward also needs.
- **#100** determines whether `masked_fill` and `where_cond` are reachable from statically shaped tensors at all. A CUDA implementation of an operation the typed frontend cannot call is still worth having, but the two land together to be useful.
- **#83** should include the out-of-range index cases as explicit rows, since a refusal is the correct behavior and a harness that only checks happy paths would not notice a kernel reading past the end.
- **#106** owns `conv1d` and `conv_transpose2d`, which share the geometry descriptor `unfold` must reuse.
- **`docs/capabilities.md`** regenerates from the registration.

## Alternatives

Restricting the index operations to contiguous inputs would simplify the kernels, but `gather` and `index_select` are most often applied to a transposed or sliced view, which is exactly the case that would then fall back.

## Additional context

Relevant code: `crates/incin-backends/src/cuda/ops/select.rs`, `crates/incin-backends/src/cuda/ops/shape.rs`, `crates/incin-backends/src/cuda/backend/shape_ops.rs`, `docs/OPERATION_SEMANTICS.md`, and the CPU references under `crates/incin-backends/src/cpu/ops/`.

Prerequisite for the attention modules and the mixture-of-experts work in this milestone.

**Comment by xupremix** [comment]
Rechecked against `master`. **All 11 operations named here are now advertised on CUDA** — the five indexing rows and the six shape/layout rows, verified individually.

The acceptance criteria are unmet and, unusually for this set, most of them cannot be satisfied by a capability row at all. "An out-of-range index produces the documented typed error and never reads outside the allocation", "`embedding` backward accumulates repeated indices correctly", and "non-contiguous, wrong-dtype, wrong-device inputs remain rejected before launch" are all statements about behaviour under inputs a declaration says nothing about.

That makes this the issue in the CUDA set where advertised-versus-verified matters most: these are the operations where a wrong index is a memory-safety question rather than a numerical one, as the problem statement says. A row asserting `gather` runs on CUDA is not evidence that an out-of-range index is caught.

Blocked on #83 for the same reason as its siblings, but with a sharper consequence if the assumption is wrong.
